proposals/0016-simd-application-fees.md

```diff
@@ -0,0 +1,278 @@
+It is also more easier to calculate total fees.
+
+### A new application fee program
+
+We add a new native solana program called application fees program with program id.
+
+```
+App1icationFees1111111111111111111111111111
+```
+
+This native program will be used to check application fee for an account, to intialize rebates
+initaited by the account authority, and a special instruction by which fee payer accepts
+amount of lamports they are willing to pay as application fees per account.
+
+#### PayApplicationFees Instruction
+
+With this instruction, the fee payer accepts to pay application fees specifying the amount.
+This instruction **MUST** be included in the transaction that interacts with dapps having application fees.
+This instruction is like an irrevokable transfer if the payer has enough funds
+i.e even if the transaction fails the payer will end up paying.
+If this instruction is added then even if the dapp does not check the application fees the payer ends
+up paying.
+If the payer does not have enough balance the transaction fails with error `InsufficientFunds`.
+
+It requires:
+
+Accounts :
+* List of accounts that need an application fees.
+
+Argument : Corresponding list of application fees for each account in `Accounts`.
+
+The arguments list must be of same length as number of accounts.
+The index of fees and account should match.
+The account fees for each account is 8 bytes represented in rust type `u64`
+
+#### CheckApplicationFees Instruction
+
+This instruction will check if an application fee for an account is paid.
+It requires :
+
+* Account where fees are paid.
+
+Argument: required fees in lamport (u64).
+
+If application fees are not paid or are paid insufficiently this instruction will return
+an error. The idea is dapp developer uses this instruction to check if the required fees
+are paid and fail the transaction if they are not paid or partially paid.
+In case of partial payment, the user will lose the partially paid amount.
+A payer may overpay for the fees. This instruction can be called multiple times across multiple instructions.
+
+#### Rebate Instruction
+
+This instruction should be called by the dapp using CPI or by the owner of the account.
+It requires :
+
+* Account on which a fee was paid
+* Authority (owner) of the account (signer)
+
+Argument: Number of lamports to rebate (u64) can be u64::MAX to rebate all the fees.
+
+The authority or the owner could be easily deduced from the `AccountMeta`. In case of PDA's
+usually account and owner are the same (if it was not changed), then `invoke_signed` can be used
+to issue a rebate.
+In case of multiple rebate instructions, only the maximum rebate will one will be issued.
+Payer has to pay full application fees even if they are eligible for a rebate.
+If there is no application fee associated with the account we ignore the rebate instruction. 
+
+### Changes in the core solana code
```

Comment:
Is it correct if I interpret the proposed changes to solana core this way: 
1. L1 recognizes access to dapp accounts are better to be managed by dapp devs instead of at network level, as dapps devs have better insight, and incentive to do so. 
2. Therefore L1 provides an api by honoring `PayApplicationFee` IX to transfer lamports and verifying payer balance before locking accounts. 
3. The rest are to be dealt directly between dapp users and devs.

Reply:
It is more like
1) L1 has to honor `PayApplicationFee` Ix as you said.
2) L1 has to track the application fees and rebates by a special structure in invoke context.
3) At the end of batching if the transaction is executed successfully we rebate the amount in rebates to the payer and transfer the application fees to corresponding accounts. If the transaction fails the we just transfer application fees to the respective accounts.
4) Checking if application fees are paid and issuing rebates are done on dapp side.